tests/app/questionnaire/test_questionnaire_manager.py

```diff
@@ -0,0 +1,145 @@
+
+        question = MagicMock()
+        question.children = MagicMock()
+        self.questionnaire_manager.state.find_state_item = MagicMock(return_value=question)
+
+        post_data = {
+            'action[add_answer]': '',
+        }
+
+        with patch.object(self.questionnaire_manager, 'update_questionnaire_store') as mock:
+            self.questionnaire_manager.add_answer('block', 'question', post_data, self.answer_store)
+            mock.assert_called_with('block')
+
+    def test_remove_answer_detaches_answer_from_question(self):
+        post_data = {
+            'action[remove_answer]': '1',
+        }
+
+        answer_to_remove = self.answers[1]
+
+        with patch.object(self.question, 'remove_answer') as mock:
+            self.questionnaire_manager.remove_answer('block', post_data, self.answer_store)
+            mock.assert_called_with(answer_to_remove)
+
+    def test_remove_answer_removes_from_answer_store(self):
+        post_data = {
+            'action[remove_answer]': '1',
+        }
+
+        answer_to_remove = self.answers[1]
+        flattened_answer = answer_to_remove.flatten()
+
+        with patch.object(self.answer_store, 'remove') as mock:
+            self.questionnaire_manager.remove_answer('block', post_data, self.answer_store)
+            mock.assert_called_with(flattened_answer)
+
+    def test_create_new_answer_no_previous_answers(self):
+        # Given
+        answer_schema = MagicMock()
+        answer_schema.widget.name = 'answer'
+
+        answer_state = MagicMock()
+        answer_state.schema_item = answer_schema
+
+        answer_schema.construct_state = MagicMock(return_value=answer_state)
+        question_state = MagicMock()
+
+
+        # When
+        new_answer = self.questionnaire_manager._create_new_answer(answer_schema, question_state, self.answer_store)
+
+        # Then
+        self.assertEqual(new_answer.schema_item.widget.name, 'answer_0')
+        self.assertEqual(new_answer.instance, 0)
+
+    def test_create_new_answer_with_existing_answers(self):
+        # Given
+        answer_schema = MagicMock()
+        answer_schema.widget.name = 'answer'
+
+        answer_state = MagicMock()
+        answer_state.schema_item = answer_schema
+
+        answer_schema.construct_state = MagicMock(return_value=answer_state)
+        question_state = MagicMock()
+
+        existing_answers = [{
+            'answer_id': 'answer',
+            'answer_instance': '0',
+        },
+        {
+            'answer_id': 'answer',
+            'answer_instance': '1',
+        }]
+
+        self.answer_store.filter = MagicMock(return_value=existing_answers)
+        # When
+        new_answer = self.questionnaire_manager._create_new_answer(answer_schema, question_state, self.answer_store)
+
+        # Then
+        self.assertEqual(new_answer.schema_item.widget.name, 'answer_2')
+        self.assertEqual(new_answer.instance, 2)
+
+    def setUp(self):
```

Comment:
Minor point but it's a bit confusing reading the tests with this at the bottom as you have references to things you don't expect to exist but that might be a preference thing